ChainSafe/lodestar
# feat: add `debug_getHistoricalSummaries` endpoint

**Motivation**

The Portal Network needs a way to access the `HistoricalSummaries` property on the `BeaconState` for post Capella beacon state since the HistoricalSummaries are used to anchor EL block header proofs to the beacon state roots in light client updates that the serve as the source of truth for verifying the authenticity of post capella EL block headers.  This currently requires downloading the entire Beacon State from a beacon node willing to serve the `debug` endpoints and then throwing away everything except for the desired data (or else downloading an ERA file from a willing endpoint).  It is desirable to be able to access this data more efficiently directly from beacon nodes.  

**Description**

This adds a new `getHistoricalSummaries` endpoint to the `debug` namespace on the REST API that serves the `HistoricalSummaries` field from the current `head` BeaconState as SSZ bytes.  We only need the most current one for use in constructing proofs of recent EL block headers embedded in a beacon block.

This mirrors similar [experimental work being done in the nimbus client](https://github.com/status-im/nimbus-eth2/pull/6675)

**Steps to test or reproduce**

```sh
git checkout historicalSummaries-endpoint
./lodestar beacon \
--rest.namespace='*' \
--checkpointSyncUrl=https://beaconstate.info \
```

In a separate terminal
```sh
curl -X GET --header "Accept: application/octet-stream" "http://127.0.0.1:9596/eth/v1/debug/historical_summaries" > ./hs.ssz
```

Then, use a script like below to verify that historical summaries are correctly returned
```ts
    import { ssz } from '@lodestar/types'

    const summariesBytes = fs.readFileSync('./test/util/hs.ssz')
    const summaries = ssz.capella.BeaconState.fields.historicalSummaries.deserialize(summariesBytes)
    console.log(summaries)
```
